Fix character volume calculation

feinorgh opened the pull request:
#### Summary
Balance "Bring volume calculations for player body more in line with realistic examples"

#### Purpose of change
After the addition of volume calculations in #74162, my current character could no longer use any vehicles. The change is well meaning and brings realism, which I like, but the values seemed off, so I set out to make reasonably sized characters (both height and weight wise) able to get into car seats again, while retaining the limits for huge characters.

#### Describe the solution
I looked up how to calculate volume in the real world, from body density, which itself is dependent on the body fat to muscle and bone ratio, and age (age is NOT included in these calculations currently).

I tried calculating the volume sizes for various heights and sizes to see if they seem reasonable, the results can be found here: #74263.

I changed the volume calculation method for the character, and adjusted the unit tests accordingly.

The previous tests now pass. I removed some assertions based on assumptions on a default character having exactly 70 000 ml volume at 175 cm height, but the functional tests for getting into a vehicle and being cramped now works.

I play tested with my current character too (200 cm tall, normal weight), and she can now get into a car seat, even though she gets the "cramped" debuff.

#### Describe alternatives you've considered
Another approach would be to change the way car seats and volume calculations are performed, but this seemed like a MUCH bigger task, which would bring little benefit to the game or the maintainability of vehicles and vehicle parts.

#### Testing
The tests for the volume calculations and getting into a vehicle works, and gives the same results as before.

I've play tested the change, and now my character has to take off her rucksack to be able to sit in the driver seat, but she can drive the car, even though it's cramped.

#### Additional context
![Tawnya Ivey _2024-06-06T13-10-24](https://github.com/CleverRaven/Cataclysm-DDA/assets/3132980/dcdd6ff2-b8d8-4582-936a-f97d4ca7cfeb)
With military rucksack on.

![Tawnya Ivey _2024-06-06T13-11-02](https://github.com/CleverRaven/Cataclysm-DDA/assets/3132980/330fd987-e790-4c3c-b501-0abad89e8f2b)
With military rucksack taken off.

![Tawnya Ivey _2024-06-06T13-08-47](https://github.com/CleverRaven/Cataclysm-DDA/assets/3132980/f7314cca-bc52-40e2-9d69-8643c6ce5354)
Driving nude, in the middle of summer.

<!--README: Cataclysm: Dark Days Ahead is released under the Creative Commons Attribution ShareAlike 3.0 license.
The code and content of the game are free to use, modify, and redistribute for any purpose whatsoever.
By contributing to the project you agree to the terms of the license and that any contribution you make will also be covered by the same license.
See http://creativecommons.org/licenses/by-sa/3.0/ for details. -->

feinorgh committed:
Provide body volume calculation based on body density

feinorgh committed:
Simplify calculations

feinorgh committed:
Adjust test cases to the volume calculation

feinorgh committed:
Adjust volumes for baseline tests too

IdleSol commented:
There's a problem with bionics in this method. If I understand correctly, we are using the average human body density.

But by installing the module, we increase the mass of the character. Or we change his density, since installing a module does not change the volume or height of the character. The formula does not take this into account.

Just for reference, installing all bionics will give a minimum of +60 kg. (Quick test with scales: naked man 77 kg, the same with all bionics: error weight more than 150 kg).

feinorgh commented:
> There's a problem with bionics in this method. If I understand correctly, we are using the average human body density.
> 
> But by installing the module, we increase the mass of the character. Or we change his density, since installing a module does not change the volume or height of the character. The formula does not take this into account.
> 
> Just for reference, installing all bionics will give a minimum of +60 kg. (Quick test with scales: naked man 77 kg, the same with all bionics: "error: weight more than 150 kg").

Ah, I understand. The formula assumes fat, bone and muscle. I suppose we can subtract the weight of the installed bionics from the total weight and use that?

I'll give it a try and see what I can come up with.

feinorgh commented:
> There's a problem with bionics in this method. If I understand correctly, we are using the average human body density.
> 
> But by installing the module, we increase the mass of the character. Or we change his density, since installing a module does not change the volume or height of the character. The formula does not take this into account.
> 
> Just for reference, installing all bionics will give a minimum of +60 kg. (Quick test with scales: naked man 77 kg, the same with all bionics: "error: weight more than 150 kg").

At a closer look, bodyweight() returns the current weight of the character without bionics. However, when your character is standing on a scale, their weight is calculated by get_weight(), which sums bodyweight(), inventory weight, weight of worn items, weight of wielded items, and the weight of installed bionics. So I think the original calculation is correct, bodyweight() returns only the base weight of the character, and the density calculation holds.

feinorgh commented:
I did some deeper research into body volume calculations, the results of which can be found here: https://github.com/feinorgh/body_volume_calculation

I have not compared the results of the implementation in C++ with the Python variants, further adjustments might be needed to fine tune it.

Anyway, hopefully this helps with estimating the usefulness of the PR.

feinorgh commented:
I accidentally pushed a rebase onto master, had to reset and force push to bring it back in line, but tags remain from the commits that weren't meant to be pushed.

RenechCDDA commented on tests/char_volume_test.cpp:
```diff
@@ -44,19 +44,18 @@ TEST_CASE( "character_baseline_volumes", "[volume]" )
     Character &you = get_player_character();
     REQUIRE( you.get_mutations().empty() );
     REQUIRE( you.height() == 175 );
-    CHECK( you.get_base_volume() == 70_liter );
```
There should still be a check for baseline volume to make sure it doesn't get too wild. If need be, you can force certain values (e.g. the body fat %), as that would still provide useful coverage over *the rest of the values*.

feinorgh replied:
Good point. I'll make some tests to make sure height, weight, and volume is within reasonable bounds.

RenechCDDA commented on tests/char_volume_test.cpp:
```diff
@@ -44,19 +44,18 @@ TEST_CASE( "character_baseline_volumes", "[volume]" )
 
     REQUIRE( your_height_with_trait( trait_HUGE ) == 280 );
-    CHECK( your_volume_with_trait( trait_HUGE ) == 286720_ml );
+    CHECK( your_volume_with_trait( trait_HUGE ) == 156228_ml );
```
I'm concerned that this is further separating character volume from monster volume by quite a bit. The *minimum* volume to a monster to be HUGE is 483.76L. After this, it seems the **average** HUGE character volume is merely 156.23L, or about a third of volume.

This isn't a blocker, but I think it's worth pointing out.

feinorgh replied:
I see. I'm not familiar with how the HUGE trait is meant to work. Physics wise, assuming similar densites around 1kg/L, a HUGE monster (483L) would weigh around 500 kg too. Depending on how realistic we want to be, we could also adjust densities depending on the monster in the future (say, a bloated zombie filled with gas would be less dense than a kevlar hulk).
Game play wise, it would be nice to make monsters and the character behave similarly with the same physical constrains where possible.

feinorgh committed:
Make checks for volume pass within reasaonable ranges

Relying on volumes down to the ml allows us to make changes to volume
calculations, and keep the tests withing reasaonable boundaries.

references #74263

RenechCDDA commented on tests/char_volume_test.cpp:
```diff
@@ -44,18 +44,23 @@ TEST_CASE( "character_baseline_volumes", "[volume]" )
     REQUIRE( your_height_with_trait( trait_SMALL2 ) == 70 );
-    CHECK( your_volume_with_trait( trait_SMALL2 ) == 23326_ml );
+    CHECK( ( your_volume_with_trait( trait_SMALL2 ) > 20_liter &&
+             your_volume_with_trait( trait_SMALL2 ) < 30_liter ) );
```
I don't see any reason to check within a range, any changes to volume should always be caught by this test. Anyone updating the volume calculations should also be capable of updating the test values - if not, *they really shouldn't be changing the volume calculations*.

feinorgh replied:
I'll fix that :-)

feinorgh committed:
Make checks for volume less dependent on exact volumes

Make tests more flexible

feinorgh committed:
Make tests exact, set character to exact calorie count

This strategy makes the test repeatable

feinorgh committed:
Remove unused header units.h

RenechCDDA reviewed (approved):
Any concerns I've mentioned have been addressed. The Brozek formula is a more accurate way to measure human body volume. Tests have been properly updated to account for the new formula.
